src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,900 @@
+            {
+                throw new IOException("No SPI response");
+            }
+            else
+            {
+                for (int ndx = 4; ndx < 24; ++ndx)
+                {
+                    if (reply[ndx] != 0)
+                    {
+                        retVal += (char)reply[ndx];
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+            }
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read the 20 characters of GoPiGo3 board name
+        /// </summary>
+        /// <returns>Returns the GoPiGo3 board name string</returns>
+        public string GetBoard()
+        {
+            string retVal = string.Empty;
+            byte[] outArray = {SpiAddress, (byte)SpiMessageType.GetName,
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == 0xA5)
+            {
+                for (int ndx = 4; ndx < 24; ++ndx)
+                {
+                    if (reply[ndx] != 0)
+                    {
+                        retVal += (char)(reply[ndx]);
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+            }
+            else
+            {
+                throw new IOException("No SPI response");
+            }
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read the hardware version
+        /// </summary>
+        /// <returns>Returns the hardware version as a string</returns>
+        public string GetHardwareVersion()
+        {
+            string retVal = string.Empty;
+            int version = SpiRead32(SpiMessageType.GetHardwareVersion);
+            retVal = $"{version / 1000000}.{(version / 1000) % 1000}.{version % 1000}";
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read the 128 bit GoPiGo3 hardware serial number
+        /// </summary>
+        /// <returns>Returns the serial number as 32 char HEX formatted string</returns>
+        public string GetId()
+        {
+            string retVal = string.Empty;
+            byte[] outArray = {SpiAddress, (byte)SpiMessageType.GetId,
+                     0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == 0xA5)
+            {
+                retVal = string.Join("", reply.Skip(4).Take(16).Select((b) => b.ToString("X2")));
+            }
+            else
+            {
+                throw new IOException("No SPI response");
+            }
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read the firmware version
+        /// </summary>
+        /// <returns>Returns the firmware version</returns>
+        public string GetFirmwareVersion()
```

Comment:
I wonder if [Version](https://docs.microsoft.com/en-us/dotnet/api/system.version?view=netcore-2.1) would suit here

Reply:
yes, implementing this Version. Makes sense.